llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp

```diff
@@ -997,8 +997,17 @@ AArch64LoadStoreOpt::mergePairedInsns(MachineBasicBlock::iterator I,
       //   STRWui %w0, ...
       //   USE %w1
       //   STRWui kill %w1  ; need to clear kill flag when moving STRWui upwards
-      RegOp0.setIsKill(false);
-      RegOp1.setIsKill(false);
+      for (auto It = std::next(I);
+           It != Paired && (RegOp0.isKill() || RegOp1.isKill()); ++It) {
+        auto ClearKill = [](MachineInstr &MI, MachineOperand &MOP,
+                            const TargetRegisterInfo *TRI) {
+          Register Reg = MOP.getReg();
+          if (MI.readsRegister(Reg, TRI))
+            MOP.setIsKill(false);
+        };
+        ClearKill(*It, RegOp0, TRI);
```

Comment:
@MatzeB I am wondering if clearing flags for both `RegOp0` and `RegOp1` is necessary. Take the sample you wrote above for example, the merge should not change the flags on `w0` because the read/write order of `w0` is not modified. Correct me if I am wrong.

Reply:
It's been a while since I worked on this, so I hope I am not missing anything. But intuitively you would only need to change the flags for operands that you move around the schedule (as in conceptually you are moving one store next to the other store before replacing it with the `strd`).